File: backend/tests/authenticationTests/loginModuleTests.py
import pytest
import requests

# Base URL of the backend server
BASE_URL = "http://localhost:8000/auth"

# Test data
VALID_USER = {
    "username": "admin",
    "password": "admin"
}

INVALID_USER = {
    "username": "invaliduser",
    "password": "wrongpassword"
}

@pytest.fixture
def setup_test_user():
    """
    Fixture to set up a test user in the database.
    This assumes you have an endpoint to create users for testing purposes.
    """
    # Create a test user
    response = requests.post(f"{BASE_URL}/register", json=VALID_USER)
    assert response.status_code == 201  # Ensure the user is created successfully
    yield
    # Clean up the test user
    requests.delete(f"{BASE_URL}/delete", json={"username": VALID_USER["username"]})


def test_login_success(setup_test_user):
    """
    Test successful login with valid credentials.
    """
    response = requests.post(f"{BASE_URL}/login", json=VALID_USER)
    assert response.status_code == 200
    assert response.json()["data"]["type"] == "Success"
    assert response.json()["data"]["message"] == "Login successfuly"


def test_login_invalid_credentials():
    """
    Test login with invalid credentials.
    """
    response = requests.post(f"{BASE_URL}/login", json=INVALID_USER)
    assert response.status_code == 400
    assert response.json()["error"]["type"] == "BadRequest"
    assert response.json()["error"]["message"] == "Invalid credentials"


def test_login_missing_fields():
    """
    Test login with missing fields in the request body.
    """
    response = requests.post(f"{BASE_URL}/login", json={"username": "testuser"})
    assert response.status_code == 400
    assert response.json()["error"]["type"] == "BadRequest"


def test_login_server_error(monkeypatch):
    """
    Test login when a server error occurs.
    """
    def mock_find_one(*args, **kwargs):
        raise Exception("Database error")

    # Mock the database call to simulate a server error
    monkeypatch.setattr("backend.authentication.authenticationModel.User.find_one", mock_find_one)

    response = requests.post(f"{BASE_URL}/login", json=VALID_USER)
    assert response.status_code == 500
    assert response.json()["error"]["type"] == "InternalServerError"
    assert response.json()["error"]["message"] == "Server error"

File: backend/tests/authenticationTests/loginModuleTests.js
const login = require('../backend/authentication/authenticationModules/loginModule');
const User = require('../backend/authentication/authenticationModel');
const { badRequestJsonResponse, internalErrorJsonResponse, successJsonResponse } = require('../backend/utils/jsonResponses/jsonResponses');

jest.mock('../backend/authentication/authenticationModel'); // Mock the User model

describe('Login Module', () => {
  let req, res;

  beforeEach(() => {
    // Mock request and response objects
    req = {
      body: {
        username: '',
        password: '',
      },
    };
    res = {
      status: jest.fn().mockReturnThis(),
      json: jest.fn(),
    };
  });

  it('should return 400 if username or password is invalid', async () => {
    // Mock User.findOne to return null (user not found)
    User.findOne.mockResolvedValue(null);

    req.body.username = 'invaliduser';
    req.body.password = 'wrongpassword';

    await login(req, res);

    expect(res.status).toHaveBeenCalledWith(400);
    expect(res.json).toHaveBeenCalledWith(badRequestJsonResponse('Invalid credentials'));
  });

  it('should return 400 if password does not match', async () => {
    // Mock User.findOne to return a user with a different password
    User.findOne.mockResolvedValue({ username: 'testuser', password: 'correctpassword' });

    req.body.username = 'testuser';
    req.body.password = 'wrongpassword';

    await login(req, res);

    expect(res.status).toHaveBeenCalledWith(400);
    expect(res.json).toHaveBeenCalledWith(badRequestJsonResponse('Invalid credentials'));
  });

  it('should return 200 if username and password are valid', async () => {
    // Mock User.findOne to return a user with matching credentials
    User.findOne.mockResolvedValue({ username: 'testuser', password: 'testpassword' });

    req.body.username = 'testuser';
    req.body.password = 'testpassword';

    await login(req, res);

    expect(res.status).toHaveBeenCalledWith(200);
    expect(res.json).toHaveBeenCalledWith(successJsonResponse('Login successfuly'));
  });

  it('should return 500 if a server error occurs', async () => {
    // Mock User.findOne to throw an error
    User.findOne.mockRejectedValue(new Error('Database error'));

    req.body.username = 'testuser';
    req.body.password = 'testpassword';

    await login(req, res);

    expect(res.status).toHaveBeenCalledWith(500);
    expect(res.json).toHaveBeenCalledWith(internalErrorJsonResponse('Server error'));
  });
});